### ISSUE TITLE
Discussion: Removal or rebase of multi-revision support

### ISSUE THREAD

**Issue by ryleykimmel** [apollo-rsps/apollo#79]
Apollo adds a lot of unnecessary (and frankly confusing to some) bloat with the current hack-job implementation of multi-revision support. I say hack-job as it is not _actual_ multi-revision support which people think of (running 1 server, connecting to it via multiple revision clients).

Don't get me wrong, Apollo's multi-revision idea is nice,... but it isn't ideal.

There are several alternative solutions that can be brought to the table that will work much nicely.

- The most simple, drop multi-revision support entirely and emulate one revision
    - Improves code flow and readability
    - No (more) issues when writing content for 377 that (will ultimately have to exist in the 317 revision as well) won't work because of compatibility issues. Even by 'disabling' content for certain revisions that aren't compatible will add an unnecessary layer of complexity to content code -- which we don't want.
    - Limits the scope to where certain bugs/issues may occur (ahem the sync [or not sync] related movement issue)
    - However... we also run into some issues with this design. Designing the server around one revision ultimately makes the core lack certain functions and necessities to let other revisions work (if someone were to convert the stock revision to another)

- Create multiple revision independent Apollo branches (or versions)
    - Not ideal for developers

- Or the best solution, in my opinion. Add support for loading 3rd party revisions without the need to hack anything on to the core
    - Abstract the core and provide it with the necessities to support a wide array of potential revisions
    - Similar to the way content and core separation was intended to be done, dragging and dropping revisions in as a dependency or creating some kind of Apollo package manager for releases (for their content and release specific code)

Discuss, ask questions, propose your own solution or call me an idiot, doesn't matter!

**Comment by Major-** [comment]
I don't like the idea of removing the existing abstraction, but bundling the 317 release separately is something I am quite keen on.

**Comment by ryleykimmel** [comment]
The existing abstraction isn't really something I like to work with either. Option 2 seems to be something that would be more ideal if you were to bundle different releases.

Abstracting some core parts of the server that are necessary for revision specific code would work nicely (keep one API for each release branch, that way fixes to the core won't have to be duplicated) and simply branch each release.

**Comment by WizardJesse1** [comment]
I never understood the whole reason for the multi-revision in apollo since you need to change the login decoder and etc.. to get a different protocol working. but I do like this since I don't have to make 20 different apollo sources for one revision since I use apollo as a 474 and etc.

**Comment by lare96** [comment]
I agree, not a fan of the bloat that comes with supporting multiple revisions.

**Comment by Major-** [comment]
@WizardJesse1 apollo is only designed for 3xx anyway

**Comment by WizardJesse1** [comment]
@Major- True but you know how people are

**Comment by Major-** [comment]
377 will be removed when #161 lands. The message encoders/decoders will still be available, but won't be distributed with the official release (and all plugins that ship with apollo by default will work on plain 317).

**Comment by ryleykimmel** [comment]
A step in the right direction I think

**Comment by Major-** [comment]
Considering removing 317 instead of 377. We have an actual 377 cache, whereas we don't for 317. The 317 client everyone uses also has some bugs with head/hint icons, although I haven't investigated to see if the issues are present in the obfuscated client too.

Apollo currently requires a cache from somewhere around 320-330, because some of the NPCs used in one of the location plugins don't exist in 317. Obviously we could remove these, but personally I think defaulting to 377 is the better choice (it's quite nice to be able to have 'everything' prior to 400 imo).

We'll probably stick the codec stuff for the removed revision in a separate repo

**Comment by ryleykimmel** [comment]
I think we should do that as well as construct a wiki for all of our 317 and 377 findings and then remove all of the boilerplate code out of Apollo for supporting multiple releases.
